sdk/python/v1beta1/kubeflow/katib/api/katib_client.py

```diff
@@ -302,137 +424,347 @@ def tune(
-                    models.V1beta1TrialParameterSpec(name=p_name, reference=p_name)
+        # If users choose to use a custom objective function.
+        if objective is not None:
+            # Validate objective function.
+            utils.validate_objective_function(objective)
+
+            # Extract objective function implementation.
+            objective_code = inspect.getsource(objective)
+
+            # Objective function might be defined in some indented scope
+            # (e.g. in another function). We need to dedent the function code.
+            objective_code = textwrap.dedent(objective_code)
+
+            # Iterate over input parameters.
+            input_params = {}
+            experiment_params = []
+            trial_params = []
+            for p_name, p_value in parameters.items():
+                # If input parameter value is Katib Experiment parameter sample.
+                if isinstance(p_value, models.V1beta1ParameterSpec):
+                    # Wrap value for the function input.
+                    input_params[p_name] = f"${{trialParameters.{p_name}}}"
+
+                    # Add value to the Katib Experiment parameters.
+                    p_value.name = p_name
+                    experiment_params.append(p_value)
+
+                    # Add value to the Katib Experiment's Trial parameters.
+                    trial_params.append(
+                        models.V1beta1TrialParameterSpec(name=p_name, reference=p_name)
+                    )
+                else:
+                    # Otherwise, add value to the function input.
+                    input_params[p_name] = p_value
+
+            # Wrap objective function to execute it from the file. For example:
+            # def objective(parameters):
+            #     print(f'Parameters are {parameters}')
+            # objective({
+            #     'lr': '${trialParameters.lr}',
+            #     'epochs': '${trialParameters.epochs}',
+            #     'is_dist': False
+            # })
+            objective_code = f"{objective_code}\n{objective.__name__}({input_params})\n"
+
+            # Prepare execute script template.
+            exec_script = textwrap.dedent(
+                """
+                program_path=$(mktemp -d)
+                read -r -d '' SCRIPT << EOM\n
+                {objective_code}
+                EOM
+                printf "%s" "$SCRIPT" > $program_path/ephemeral_objective.py
+                python3 -u $program_path/ephemeral_objective.py"""
+            )
+
+            # Add objective code to the execute script.
+            exec_script = exec_script.format(objective_code=objective_code)
+
+            # Install Python packages if that is required.
+            if packages_to_install is not None:
+                exec_script = (
+                    utils.get_script_for_python_packages(
+                        packages_to_install, pip_index_url
+                    )
+                    + exec_script
```

Comment:
Got it! I agree that it will make the code clearer. I'll implement it as soon as possible.

Reply:
@andreyvelich I have added the helper function `get_exec_script_from_objective` in utils.py:
https://github.com/kubeflow/katib/blob/c2df96757ce2fb39b5debf60ad7ed1873ce31ae5/sdk/python/v1beta1/kubeflow/katib/utils/utils.py#L212
And I have updated the `parameter_substitution` function so that it can also be reused with custom objective functions:
https://github.com/kubeflow/katib/blob/c2df96757ce2fb39b5debf60ad7ed1873ce31ae5/sdk/python/v1beta1/kubeflow/katib/utils/utils.py#L146
Please review when you have time and let me know if you have any further suggestions.